File: src/harness_core/dispatch/dispatch_ledger.py
"""Dispatch ledger: in-memory store for DispatchRecords."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any

# ---------------------------------------------------------------------------
# Schema version
# ---------------------------------------------------------------------------

DISPATCH_RECORD_SCHEMA_VERSION = "dispatch_record.v1"

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

DISPATCH_STATUSES: tuple[str, ...] = (
    "dispatched", "executing", "completed", "failed", "escalated", "cancelled", "not_executed",
)


# ---------------------------------------------------------------------------
# Schema
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class DispatchRecord:
    dispatch_id: str
    request_snapshot: str
    task_analysis_id: str
    decision_id: str
    final_status: str  # from DISPATCH_STATUSES
    created_at: str
    updated_at: str
    execution_result_id: str | None = None
    evaluation_result_id: str | None = None
    usage_ledger_row_id: str | None = None
    budget_reservation_id: str | None = None
    schema_version: str = DISPATCH_RECORD_SCHEMA_VERSION

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "dispatch_id": self.dispatch_id,
            "request_snapshot": self.request_snapshot,
            "task_analysis_id": self.task_analysis_id,
            "decision_id": self.decision_id,
            "execution_result_id": self.execution_result_id,
            "evaluation_result_id": self.evaluation_result_id,
            "usage_ledger_row_id": self.usage_ledger_row_id,
            "budget_reservation_id": self.budget_reservation_id,
            "final_status": self.final_status,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
        }


# ---------------------------------------------------------------------------
# Ledger
# ---------------------------------------------------------------------------


class DispatchLedger:
    """In-memory dispatch record store."""

    def __init__(self) -> None:
        self._records: dict[str, DispatchRecord] = {}

    def create_record(
        self,
        dispatch_id: str,
        request_snapshot: str,
        task_analysis_id: str,
        decision_id: str,
        budget_reservation_id: str | None = None,
    ) -> DispatchRecord:
        now = datetime.now(timezone.utc).isoformat()
        record = DispatchRecord(
            dispatch_id=dispatch_id,
            request_snapshot=request_snapshot,
            task_analysis_id=task_analysis_id,
            decision_id=decision_id,
            budget_reservation_id=budget_reservation_id,
            final_status="dispatched",
            created_at=now,
            updated_at=now,
        )
        self._records[dispatch_id] = record
        return record

    def update_record(
        self,
        record: DispatchRecord,
        final_status: str | None = None,
        execution_result_id: str | None = None,
        evaluation_result_id: str | None = None,
        usage_ledger_row_id: str | None = None,
    ) -> DispatchRecord:
        updated = DispatchRecord(
            dispatch_id=record.dispatch_id,
            request_snapshot=record.request_snapshot,
            task_analysis_id=record.task_analysis_id,
            decision_id=record.decision_id,
            final_status=final_status or record.final_status,
            created_at=record.created_at,
            updated_at=datetime.now(timezone.utc).isoformat(),
            execution_result_id=execution_result_id or record.execution_result_id,
            evaluation_result_id=evaluation_result_id or record.evaluation_result_id,
            usage_ledger_row_id=usage_ledger_row_id or record.usage_ledger_row_id,
            budget_reservation_id=record.budget_reservation_id,
        )
        self._records[record.dispatch_id] = updated
        return updated

    def get_record(self, dispatch_id: str) -> DispatchRecord | None:
        return self._records.get(dispatch_id)

    def list_records(self) -> list[DispatchRecord]:
        return list(self._records.values())

    def replay(self, dispatch_id: str) -> DispatchRecord | None:
        return self._records.get(dispatch_id)

File: engine/src/dispatch_ledger.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::dispatch_decision::DispatchDecision;
use crate::evaluation_stub::EvaluationResult;
use crate::executor_adapter::ExecutionResult;
use crate::runtime::FixtureRuntime;
use crate::task_analyzer::TaskAnalysis;

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct DispatchRecord {
    pub schema_version: String,
    pub dispatch_id: String,
    pub request_snapshot: String,
    pub task_analysis_id: String,
    pub decision_id: String,
    pub execution_result_id: Option<String>,
    pub evaluation_result_id: Option<String>,
    pub usage_ledger_row_id: Option<String>,
    pub budget_reservation_id: Option<String>,
    pub final_status: String,
    pub created_at: String,
    pub updated_at: String,
}

impl DispatchRecord {
    pub fn to_value(&self) -> Value {
        serde_json::to_value(self).expect("DispatchRecord should serialize")
    }
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct DispatchBundle {
    pub record: Value,
    pub analysis: Value,
    pub decision: Value,
    pub execution_result: Value,
    pub evaluation_result: Value,
}

impl DispatchBundle {
    pub fn to_value(&self) -> Value {
        serde_json::to_value(self).expect("DispatchBundle should serialize")
    }
}

#[derive(Default)]
pub struct DispatchLedger;

impl DispatchLedger {
    pub fn new() -> Self {
        Self
    }

    pub fn create_record(
        &self,
        dispatch_id: &str,
        request_snapshot: &str,
        task_analysis_id: &str,
        decision_id: &str,
        budget_reservation_id: Option<String>,
        runtime: &FixtureRuntime,
    ) -> DispatchRecord {
        DispatchRecord {
            schema_version: "dispatch_record.v1".to_string(),
            dispatch_id: dispatch_id.to_string(),
            request_snapshot: request_snapshot.to_string(),
            task_analysis_id: task_analysis_id.to_string(),
            decision_id: decision_id.to_string(),
            execution_result_id: None,
            evaluation_result_id: None,
            usage_ledger_row_id: None,
            budget_reservation_id,
            final_status: "dispatched".to_string(),
            created_at: runtime.now(),
            updated_at: runtime.now(),
        }
    }

    pub fn update_record(
        &self,
        record: DispatchRecord,
        final_status: &str,
        execution_result_id: Option<String>,
        evaluation_result_id: Option<String>,
        usage_ledger_row_id: Option<String>,
        runtime: &FixtureRuntime,
    ) -> DispatchRecord {
        DispatchRecord {
            final_status: final_status.to_string(),
            execution_result_id: execution_result_id.or(record.execution_result_id),
            evaluation_result_id: evaluation_result_id.or(record.evaluation_result_id),
            usage_ledger_row_id: usage_ledger_row_id.or(record.usage_ledger_row_id),
            updated_at: runtime.now(),
            ..record
        }
    }

    pub fn store_bundle(
        &self,
        record: DispatchRecord,
        analysis: TaskAnalysis,
        decision: DispatchDecision,
        execution_result: ExecutionResult,
        evaluation_result: EvaluationResult,
    ) -> DispatchBundle {
        DispatchBundle {
            record: record.to_value(),
            analysis: analysis.to_value(),
            decision: decision.to_value(),
            execution_result: execution_result.to_value(),
            evaluation_result: evaluation_result.to_value(),
        }
    }
}
